### PR TITLE
Bce logit loss

### PR DESCRIPTION
**Pull request by mtijink**
Hi, thanks for your library! I used it with great success, getting me second place in a chess-like competition (codecup.org).

One operation I missed is "binary cross entropy with logits loss" (as PyTorch calls it). To me, that is the natural thing to use as a loss function. Mean-square-error on `sigmoid(y)` leads to flat gradients if `y` is not close to zero. Cross entropy loss means introducing a second outcome, even though that's not really necessary. And the logit part is important, since that makes the gradient well-behaved under all circumstances (vs calculating `sigmoid(y)` and then doing the BCE loss directly).

I did not use CUDA, and couldn't get Rocm/HIP to work, so this is just the CPU implementation.

### CODE REVIEW

**Comment by jw1912** [comment]
I am not merging anything on the backend side for now because it is currently being rewritten (https://github.com/jw1912/bullet/pull/488). But besides that this operation, seeing as it is only for one backend, should be put in your example file - there is no need to touch any underlying bullet code, the library is intentionally extensible.

Additionally, in order to change the loss function in all the examples you would need to provide evidence of elo gain in standard chess (as that is what the examples are for) *and* have everyone switch to using it so that it is standard enough to go in an example.

**Comment by mtijink** [comment]
I did some research, and am surprised to discover that indeed nearly everyone uses MSE loss on the sigmoid of the output! But I couldn't figure out why though. Of course it does make sense to use that in the examples, given that's the case. 

Are you open to a contribution adding this function, once your backend rewrite is done? It would be a small way of contributing back to a library that's been useful to me.

**Comment by jw1912** [comment]
Definitely! It will hopefully be much simpler to do anyways after the rewrite.
